### PR TITLE
Implement OPTIONS Request

### PR DESCRIPTION
**Pull request by christophermanahan**
### Description

- Modify Router to handle OPTIONS requests

### CODE REVIEW

File: src/main/java/io/github/christophermanahan/carnitas/Application.java [hunk 1]
```diff
@@ -17,17 +18,19 @@
 
     private Function<HTTPRequest, HTTPResponse> okHandler() {
         return (HTTPRequest request) -> new ResponseBuilder()
-          .setStatus(HTTPResponse.Status.OK)
-          .addHeader(Headers.contentLength(0))
-          .get();
+          .set(HTTPResponse.Status.OK)
+          .set(List.of(
+            Headers.contentLength(0)
```
Comment (damonkelley) [inline src/main/java/io/github/christophermanahan/carnitas/Application.java]: While I like the idea of overloading set to make the API consistent, I think `addHeader` makes more sense than requiring the user to provide a list of headers.

Also, I would think that `Headers` would be a collection.

Comment (damonkelley) [reply src/main/java/io/github/christophermanahan/carnitas/Application.java]: Further, do we need to have a special method for `contentLength`?

File: src/main/java/io/github/christophermanahan/carnitas/Route.java [hunk 2]
```diff
@@ -17,7 +17,13 @@ public String uri() {
...
-    boolean equals(Route route) {
+    public boolean equals(Object object) {
+        Route route = (Route) object;
         return method.equals(route.method) && uri.equals(route.uri);
     }
+
+    public int hashCode() {
+        int valueEquality = 0;
+        return valueEquality;
+    }
```
Comment (damonkelley) [inline src/main/java/io/github/christophermanahan/carnitas/Route.java:31]: This works? Can you elaborate on this?

File: src/main/java/io/github/christophermanahan/carnitas/Router.java [hunk 3]
```diff
@@ -15,20 +15,25 @@
...
         return this;
     }
 
     Router post(String uri, Function<HTTPRequest, HTTPResponse> handler) {
-        map.put(new Route(HTTPRequest.Method.POST, uri), handler);
+        mapRoute(HTTPRequest.Method.POST, uri, handler);
         return this;
     }
 
+    private void mapRoute(HTTPRequest.Method method, String uri, Function<HTTPRequest, HTTPResponse> handler) {
```
Comment (damonkelley) [inline src/main/java/io/github/christophermanahan/carnitas/Router.java]: What is `mapRoute`? Is `map` a verb here? It confusing because we also have the "map" data structure.

File: src/main/java/io/github/christophermanahan/carnitas/Router.java [hunk 4]
```diff
@@ -56,14 +61,27 @@ public HTTPResponse handle(HTTPRequest request) {
 
     private Supplier<HTTPResponse> not(List<HTTPRequest.Method> allowed) {
         return new ResponseBuilder()
-          .setStatus(HTTPResponse.Status.METHOD_NOT_ALLOWED)
-          .addHeader(Headers.contentLength(0))
-          .addHeader(Headers.allow(allowed));
+          .set(HTTPResponse.Status.METHOD_NOT_ALLOWED)
+          .set(List.of(
+            Headers.contentLength(0),
+            Headers.allow(allowed)
```
Comment (damonkelley) [inline src/main/java/io/github/christophermanahan/carnitas/Router.java]: Same comment here about having a dedicated `allow` method.

File: src/test/java/gradle/cucumber/Steps.java [hunk 5]
```diff
@@ -92,4 +94,12 @@ public void statusCodes(int code) {
     public void theNdResponseShouldHaveBody(int index, String body) {
         assertEquals(body, responses.get(index - 1).body());
```
Comment (damonkelley) [inline src/test/java/gradle/cucumber/Steps.java:94]: Not part of this change, but it this a typo?

File: src/main/java/io/github/christophermanahan/carnitas/Application.java (hunk 1 above)
Comment (damonkelley) [reply src/main/java/io/github/christophermanahan/carnitas/Application.java]: Or if you really like having the method be generic, you could introduce a `Header` type and add an `add` method. 

```java
public void add(Header header) {
    headers.add(header);
} 
```

File: src/main/java/io/github/christophermanahan/carnitas/HTTPResponse.java [hunk 6]
```diff
@@ -44,7 +45,11 @@ Status status() {
     public boolean equals(Object object) {
         HTTPResponse response = (HTTPResponse) object;
         return status.equals(response.status)
-          && headers.equals(response.headers)
+          && new HashSet<>(headers).equals(new HashSet<>(response.headers))
```
Comment (damonkelley) [inline src/main/java/io/github/christophermanahan/carnitas/HTTPResponse.java]: Can you talk about why you decided to use a hash set?

File: src/main/java/io/github/christophermanahan/carnitas/Route.java (hunk 2 above)
Comment (christophermanahan) [reply src/main/java/io/github/christophermanahan/carnitas/Route.java:31]: I actually don't think this should work. If 0 is used as the key in the Router HashMap for all Routes, I believe the keys would override each other. I switched it over to `(method+uri).hashCode()`

File: src/main/java/io/github/christophermanahan/carnitas/HTTPResponse.java (hunk 6 above)
Comment (christophermanahan) [reply src/main/java/io/github/christophermanahan/carnitas/HTTPResponse.java]: I used a hashSet so that the order of the headers did not matter. I think this displays that the headers should be a collection and not a list, and then this conversion in the equals method would no longer be necessary.

File: src/main/java/io/github/christophermanahan/carnitas/Route.java (hunk 2 above)
Comment (damonkelley) [reply src/main/java/io/github/christophermanahan/carnitas/Route.java:31]: I am certain they would override necessarily, but what does that mean for the performance of a lookup operation?

A "trick" that I use in Java is to just let Intellij generate the `hashCode` method for me 😄

Comment (christophermanahan) [reply src/main/java/io/github/christophermanahan/carnitas/Route.java:31]: @damonkelley I'm not entirely sure what you mean here.

File: src/main/java/io/github/christophermanahan/carnitas/Router.java (hunk 3 above)
Comment (christophermanahan) [reply src/main/java/io/github/christophermanahan/carnitas/Router.java]: @damonkelley I originally wanted to use `route`, but it would be confusing as well because it would be used as a verb here, but we also have the Route data structure.

Comment (christophermanahan) [reply src/main/java/io/github/christophermanahan/carnitas/Router.java]: I am considering `add`

File: src/test/java/gradle/cucumber/Steps.java (hunk 5 above)
Comment (christophermanahan) [reply src/test/java/gradle/cucumber/Steps.java:94]: This is how CucumberJava autogenerates the method names. It matches `"The {int}nd response should have body {string}"`

Comment (christophermanahan) [reply src/test/java/gradle/cucumber/Steps.java:94]: Switched over to `theIntResponseShouldHaveBody`

**Comment by christophermanahan** [comment]
@damonkelley I believe I have addressed the review comments.

File: src/main/java/io/github/christophermanahan/carnitas/Application.java [hunk 7]
```diff
@@ -15,22 +15,22 @@
...
-    private Function<HTTPRequest, HTTPResponse> createdHandler() {
-        return (HTTPRequest request) -> new ResponseBuilder()
-          .setStatus(HTTPResponse.Status.CREATED)
-          .addHeader(Headers.contentLength(request.body().orElse("").length()))
-          .setBody(request.body())
+    private Function<Request, Response> createdHandler() {
+        return (Request request) -> new ResponseBuilder()
+          .set(Response.Status.CREATED)
+          .set(request.body())
+          .add(Headers.CONTENT_LENGTH + request.body().orElse("").length())
```
Comment (damonkelley) [inline src/main/java/io/github/christophermanahan/carnitas/Application.java:29]: I feel like determining the content length could be handled by the server, instead of the application. Not something we need to address now, but something to think about I would say.

File: src/main/java/io/github/christophermanahan/carnitas/Request.java [hunk 8]
```diff
@@ -2,30 +2,31 @@
...
-    private HTTPRequest(Method method, String uri, Optional<String> body) {
+    private Request(Method method, String uri, Optional<String> body) {
         this.method = method;
         this.uri = uri;
         this.body = body;
     }
 
-    HTTPRequest withBody(Optional<String> body) {
-        return new HTTPRequest(method, uri, body);
+    Request withBody(Optional<String> body) {
```
Comment (damonkelley) [inline src/main/java/io/github/christophermanahan/carnitas/Request.java:28]: Now that there is a `RequestBuilder` that is building up stat, is `withBody` useful anymore? Just curious.

File: src/main/java/io/github/christophermanahan/carnitas/Router.java [hunk 9]
```diff
@@ -38,32 +43,42 @@ public HTTPResponse handle(HTTPRequest request) {
...
+          .set(Response.Status.NOT_FOUND)
+          .add((Headers.CONTENT_LENGTH + 0));
+    }
+
+    private Function<Request, Response> options() {
+        return (Request request) -> new ResponseBuilder()
+          .set(Response.Status.OK)
+          .add(Headers.ALLOW + String.join(" ", allowed(request)))
+          .add(Headers.CONTENT_LENGTH + 0)
+          .get();
```
Comment (damonkelley) [inline src/main/java/io/github/christophermanahan/carnitas/Router.java:82]: The cohesion of this class is decreasing I think. There are more and more methods that are unrelated (in a way). We might look at extracting some of these methods at some point in the future.

For instance, `options`, `notFound` and some sort of `notAllowed` handler could all be extracted to their own classes. There is also some sort of route building in there too.

File: src/main/java/io/github/christophermanahan/carnitas/Serializer.java [hunk 10]
```diff
@@ -15,15 +15,15 @@
...
     }
 
-    private String statusLine(HTTPResponse response) {
-        return HTTPResponse.VERSION + " " + response.status().code + Serializer.CRLF;
+    private String statusLine(Response response) {
+        return Response.VERSION + " " + response.status().code + Serializer.CRLF;
     }
 
-    private String format(List<String> headers) {
+    private String format(TreeSet headers) {
```
Comment (damonkelley) [inline src/main/java/io/github/christophermanahan/carnitas/Serializer.java:22]: Does this _need_ to be a `TreeSet` or could it be more generic? (i.e. `Set<String>`).

**Review by damonkelley — APPROVED** [review]
Looks good!

File: src/main/java/io/github/christophermanahan/carnitas/Router.java (hunk 9 above)
Comment (christophermanahan) [reply src/main/java/io/github/christophermanahan/carnitas/Router.java:82]: I plan on extracting the route building when I work on the redirect story. that may also be a good opportunity to extract the classes you have mentioned.

File: src/main/java/io/github/christophermanahan/carnitas/Request.java (hunk 8 above)
Comment (christophermanahan) [reply src/main/java/io/github/christophermanahan/carnitas/Request.java:28]: There is only a ResponseBuilder as of now. The RequestBuilder is going to be added upon refactoring the Parser.
